java-frontend/src/main/java/org/sonar/java/model/AbstractTypedTree.java

```diff
@@ -57,7 +57,7 @@ public void setType(Type type) {
   }
 
   public void setInferedType(Type type) {
-    Preconditions.checkState(((JavaType) this.type).isTagged(JavaType.DEFERRED));
+    Preconditions.checkState(((JavaType) this.type).isTagged(JavaType.DEFERRED) || type == this.type, "Expected to have a deferred type but got "+this.type);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)